src/_pytest/fixtures.py

```diff
@@ -894,7 +894,17 @@ def execute(self, request):
         cached_result = getattr(self, "cached_result", None)
         if cached_result is not None:
             result, cache_key, err = cached_result
-            if my_cache_key == cache_key:
+            # comparison with `==` can fail, e.g. for NumPy arrays, so we try `is`
+            # first and guard the subsequent `==` comparison
+            cache_valid = False
+            if my_cache_key is cache_key:
+                cache_valid = True
+            else:
+                try:
+                    cache_valid = bool(my_cache_key == cache_key)
+                except (TypeError, ValueError):
+                    pass
```

Comment:
this should warn that a fragile cache key was used,
pytest should at some point require parameters to be reasonably stable/hash-able
accidentally and strangely modifiable parameters have led to many  horrendous errors as  aftermath of accidential modification

Reply:
Ok, I'll add a warning.
Regarding the `TypeError` case, is that even desirable? Classes have to actively disallow `==`, if unimplemented it defaults to `is`, which would be okay. Should we consider a class that raises `TypeError` on `__eq__` as pathological? Besides, `AttributeError` would also be an option.

Reply:
Perhaps we should straight out error out instead? Warnings have a tendency to get ignored, and we should avoid letting users running test suites with problematic setups.

So I suggest we raise an appropriate error pointing to the problematic node id instead.

Reply:
Just for clarity, @nicoddemus do you mean to also error out in the NumPy array case? After thinking about it for a second, I agree that's the most reasonable solution. Besides brittleness, another reason is that every `==` operation with an array computes a new boolean array, so with large arrays as parameters, the comparisons can get quite expensive.
I suggest to include a hint in the error message to use `scope='function'` for NumPy arrays or other types where `if a == b:` can raise an exception. The docs would ideally also mention that case.

Reply:
> Just for clarity, @nicoddemus do you mean to also error out in the NumPy array case? After thinking about it for a second, I agree that's the most reasonable solution. Besides brittleness, another reason is that every == operation with an array computes a new boolean array, so with large arrays as parameters, the comparisons can get quite expensive.

Yes that's my reasoning as well: we should straight out discourage parametrization using types that don't have standard/buggy `==` operators.

> I suggest to include a hint in the error message to use scope='function' for NumPy arrays or other types where if a == b: can raise an exception. The docs would ideally also mention that case.

Not sure, this problem is not related to scoping at all, I'm afraid. We could however add a link in the exception message to a section in the docs where this is explained, probably using `numpy` arrays as example and possible workarounds.

Reply:
> Not sure, this problem is not related to scoping at all, I'm afraid.

Not directly, only in the sense that for `'function'` scope, there is no caching and thus no problem. But that assumption might not be future-proof.

Reply:
Oh you are right. But indeed, best not assume that. 👍

Reply:
key reason i proposed the warning first is that technically we currently allow "broken" parameters
if we remove support for them it breaks test-suites that are technically working

the warning is there to start the migration, a breaking release can then disallow it completely

unfortunately api bugs require breaking changes to fix them

Reply:
Thanks @RonnyPfannschmidt those are good points.

I'm still leaning towards just breaking it in the next feature release though (don't think this will affect many users), but I'm also OK about deprecating it.

Reply:
Btw, please get the final word then so @kohr-h can finish the PR.

Reply:
what if we did something more clever with `request.param` below in `def cache_key` -- I think we could probably use object identity (`id(...)`) for `request.param`

Reply:
> if we remove support for them it breaks test-suites that are technically working

Hm, with the current release test suites are already broken, so the question is whether you prefer to add a (somewhat dirty) workaround or rather keep failing, but with a better error message. Or would it make sense to re-enable a broken feature and deprecate it at the same time?

@asottile You mean replace `==` with `is` basically?

Reply:
Quick testing shows that `is` instead of `==` could work. I'll do a push and see how the test suite does.